src/Lean/Data/FuzzyMatching.lean

```diff
@@ -0,0 +1,226 @@
+/-
+Copyright (c) 2018 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Lars König
+
+The string fuzzy matching algorithm in this file is based on the algorithm
+used in LLVM with some modifications. The LLVM algorithm itself is based on VS
+code's client side filtering algorithm.
+-/
+
+namespace Lean
+namespace FuzzyMatching
+
+
+section Utils
+
+@[specialize] private def iterateLookaround (f : (Option Char × Char × Option Char) → α) (string : String) : Array α :=
+  if string.isEmpty then
+    #[]
+  else if string.length == 1 then
+    #[f (none, string.get 0, none)]
+  else Id.run <| do
+    let mut result := Array.mkEmpty string.length
+    result := result.push <| f (none, string.get 0, string.get 1)
+
+    for i in [2:string.length] do
+      result := result.push <| f (string.get (i - 2), string.get (i - 1), string.get i)
+    result.push <| f (string.get (string.length - 2), string.get (string.length - 1), none)
+
+private def containsInOrderLower (a b : String) : Bool := Id.run <| do
+  if a.isEmpty then
+    return true
+  let mut aIt := a.mkIterator
+  for i in [:b.bsize] do
+    if aIt.curr.toLower == (b.get i).toLower then
```

Comment:
Since `Iterator.curr` uses `get` as well, doesn't this mean that any characters outside the UTF-8 boundary are effectively equal to `'a'`, `'A'` and any other characters outside the boundaries here?

Reply:
Yes, but note that `next` only visits char boundaries :)

Reply:
So `next` would have to reject the move when the target is not a valid UTF-8 character? But this code increments the position by 1 for non-UTF-8 characters.

```
/* invalid UTF-8 encoded string */
return lean_box(i+1);
```
(from `lean_string_utf8_next`)